src/query/inflight_retry.h

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright (c) 2025, valkey-search contributors
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD 3-Clause
+ *
+ */
+
+#ifndef VALKEY_SEARCH_SRC_QUERY_INFLIGHT_RETRY_H_
+#define VALKEY_SEARCH_SRC_QUERY_INFLIGHT_RETRY_H_
+
+#include <memory>
+#include <vector>
+
+#include "src/index_schema.h"
+#include "src/metrics.h"
+#include "src/utils/string_interning.h"
+#include "src/valkey_search_options.h"
+#include "vmsdk/src/log.h"
+#include "vmsdk/src/utils.h"
+#include "vmsdk/src/valkey_module_api/valkey_module.h"
+
+namespace valkey_search::query {
+
+// Base class for in-flight retry contexts. Provides common interface for
+// checking conflicts and scheduling retries.
+//
+// Usage pattern:
+// 1. Create derived context with case-specific data
+// 2. Call ScheduleOnMainThread() to start the retry loop
+// 3. The timer callback calls ProcessRetry() which handles:
+//    - Cancellation check -> `OnCancelled()`
+//    - Conflict check -> schedule retry timer
+//    - No conflicts -> `OnComplete()`
+class InFlightRetryContextBase {
+ public:
+  explicit InFlightRetryContextBase(std::vector<InternedStringPtr> keys)
+      : neighbor_keys_(std::move(keys)) {}
+
+  virtual ~InFlightRetryContextBase() = default;
+  virtual void OnComplete() = 0;
+  virtual void OnCancelled() = 0;
+  virtual bool IsCancelled() const = 0;
+  virtual const std::shared_ptr<IndexSchema>& GetIndexSchema() const = 0;
+  const std::vector<InternedStringPtr>& GetNeighborKeys() const {
+    return neighbor_keys_;
+  }
+
+ private:
+  std::vector<InternedStringPtr> neighbor_keys_;
+};
+
+// Process a retry attempt. Handles cancellation, conflict checking, and
+// completion. This is the main entry point called by timer callbacks.
+template <typename RetryContext>
```

Comment:
It's great how you've condensed the number of code paths! I actually don't think you even need the template argument. You should be able to rely on polymophism:

```
void ProcessRetry(ValkeyModuleCtx* ctx, InFlightRetryContextBase* retry_ctx,
                  void (*timer_callback)(ValkeyModuleCtx*, void*),
                  const char* log_prefix);
```

Then you wouldn't need the separate `InFlightRetryCallback`, `RemoteInFlightRetryCallback`, and `LocalInFlightRetryCallback` functions. You do use them to define separate description strings used in error messages, but those can be moved into the retry classes themselves (e.g. define a virtual void `GetDesc()` in `InFlightRetryContextBase`)